common/utils.sh

```diff
@@ -66,8 +71,36 @@ check_for_sanitizer_errors() {
             msan_cnt++;
             break;
         }
+
+        # The SUMMARY line means we are at the end of the current error/warning
+        # block, so print the current (summary) line and switch the line printing
+        # off
+        print $0;
+        sanitizer_match = 0;
+        # Also, print some kind of separator (a newline for now)
+        print "\n";
     }
 
+    # Extractors
+    # Address and Memory sanitizers
+    /==[0-9]+==.+?\w+Sanitizer/ {
+        sanitizer_match = 1;
+    }
+
+    # UndefinedBehavior sanitizer
+    /[0-9]+: runtime error/ {
+        sanitizer_match = 1;
+    }
+
+    # Generic code block
+    {
+        # If we are processing a sanitizer error/warning block, print
+        # the current line
+        if (sanitizer_match)
+            print $0;
+    }
```

Comment:
As far as I know, `awk` can maintain this for us via something like
```
awk '/([0-9]+: runtime error|==[0-9]+==.+?\w+Sanitizer)/,/SUMMARY:\s+(\w+)Sanitizer/ { print }'
```
I'm not sure how well it combines with the global state kept in the "cnt_*" variables though.

Reply:
Judging by https://www.gnu.org/software/gawk/manual/html_node/Ranges.html, it's called a range pattern.

Reply:
Interesting, I knew about the range patterns but I somehow thought it's `sed`-only thing, thanks :-)

Also, it seems like it works well with the counters:

```
$ cat test.log | check_for_sanitizer_errors
==19697==ERROR: LeakSanitizer: detected memory leaks
Direct leak of 136 byte(s) in 1 object(s) allocated from:
    #0 0x7f9ea5bfcc08 in __interceptor_malloc (/lib64/libasan.so.5+0xefc08)
    #1 0x455f41 in malloc_multiply ../src/basic/alloc-util.h:96
    #2 0x4566ce in routing_policy_rule_new ../src/network/networkd-routing-policy-rule.c:21
    #3 0x45a0de in routing_policy_rule_add_internal ../src/network/networkd-routing-policy-rule.c:265
    #4 0x45ae37 in routing_policy_rule_add ../src/network/networkd-routing-policy-rule.c:317
    #5 0x461602 in routing_policy_rule_configure ../src/network/networkd-routing-policy-rule.c:591
    #6 0x50fc61 in link_request_set_routing_policy_rule ../src/network/networkd-link.c:832
    #7 0x510e6c in link_request_set_routes ../src/network/networkd-link.c:896
    #8 0x511c03 in link_check_ready ../src/network/networkd-link.c:950
    #9 0x5133f6 in address_handler ../src/network/networkd-link.c:1037
    #10 0x7f9ea4fa371c in process_reply ../src/libsystemd/sd-netlink/sd-netlink.c:347
    #11 0x7f9ea4fa452a in process_running ../src/libsystemd/sd-netlink/sd-netlink.c:414
    #12 0x7f9ea4fa49e0 in sd_netlink_process ../src/libsystemd/sd-netlink/sd-netlink.c:443
    #13 0x7f9ea4fa718e in io_callback ../src/libsystemd/sd-netlink/sd-netlink.c:712
    #14 0x7f9ea4ff72ed in source_dispatch ../src/libsystemd/sd-event/sd-event.c:2827
    #15 0x7f9ea4ffd7d2 in sd_event_dispatch ../src/libsystemd/sd-event/sd-event.c:3240
    #16 0x7f9ea4ffe71a in sd_event_run ../src/libsystemd/sd-event/sd-event.c:3297
    #17 0x7f9ea4ffea43 in sd_event_loop ../src/libsystemd/sd-event/sd-event.c:3319
    #18 0x40d79d in run ../src/network/networkd.c:114
    #19 0x40d9a3 in main ../src/network/networkd.c:121
    #20 0x7f9ea36f2412 in __libc_start_main (/lib64/libc.so.6+0x24412)
SUMMARY: AddressSanitizer: 136 byte(s) leaked in 1 allocation(s).


../src/shared/json.c:1028:38: runtime error: 4.21e+448 is outside the range of representable values of type 'long'
    #0 0x7f8ea9da8195 in json_variant_has_type /home/travis/build/systemd/systemd/build/../src/shared/json.c:1028:38
    #1 0x7f8ea9db7a1a in json_dispatch /home/travis/build/systemd/systemd/build/../src/shared/json.c:3227:30
    #2 0x7f8ea9db7bb7 in json_dispatch /home/travis/build/systemd/systemd/build/../src/shared/json.c:3251:37
    #3 0x5821f4 in oci_load /home/travis/build/systemd/systemd/build/../src/nspawn/nspawn-oci.c:2259:13
    #4 0x58186e in LLVMFuzzerTestOneInput /home/travis/build/systemd/systemd/build/../src/fuzz/fuzz-nspawn-oci.c:24:16
    #5 0x44bbaa in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:576:15
    #6 0x43fc61 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:280:6
    #7 0x4449db in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:714:9
    #8 0x46c902 in main /tmp/final/llvm.src/projects/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #9 0x7f8ea8a3e2e0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x202e0)
    #10 0x421158 in _start (/fuzzer+0x421158)
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../src/shared/json.c:1028:38 in


==19693==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x7f7b66d653fd in is_main_thread /build/build/../src/basic/process-util.c:928:13
    #1 0x7f7b66c3989c in cleanup_pools /build/build/../src/basic/hashmap.c:288:14
    #2 0x7f7b683146f5  (/lib64/ld-linux-x86-64.so.2+0xf6f5)
    #3 0x7f7b67fa7d8b  (/lib/x86_64-linux-gnu/libc.so.6+0x39d8b)
    #4 0x7f7b67fa7eb9 in exit (/lib/x86_64-linux-gnu/libc.so.6+0x39eb9)
    #5 0x7f7b67f920a1 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x240a1)
    #6 0x420ab9 in _start (/build/build/test-dlopen+0x420ab9)
  Uninitialized value was created by a heap allocation
    #0 0x42c8bd in malloc (/build/build/test-dlopen+0x42c8bd)
    #1 0x7f7b683168b7  (/lib64/ld-linux-x86-64.so.2+0x118b7)
SUMMARY: MemorySanitizer: use-of-uninitialized-value /build/build/../src/basic/process-util.c:928:13 in is_main_thread


 ____________________________________________
/ Found   3 sanitizer errors (  1 ASan,   1  \
| UBSan,   1 MSan). Looks like you need to   |
\ look at the log                            /
 --------------------------------------------
 \
  \
     __
    /  \
    |  |
    @  @
    |  |
    || |/
    || ||
    |\_/|
    \___/
```